File: backend/src/main/java/com/jbconnect/healthclinix/entity/user/adapter/GetByFiltersAdapter.java
package com.jbconnect.healthclinix.entity.user.adapter;

import com.jbconnect.healthclinix.entity.user.dto.UserRequestDTO;
import com.jbconnect.healthclinix.entity.user.dto.UserResponseDTO;
import com.jbconnect.healthclinix.entity.user.mapper.UserMapper;
import com.jbconnect.healthclinix.entity.user.model.UserModel;
import com.jbconnect.healthclinix.entity.user.port.GetByFiltersPort;
import com.jbconnect.healthclinix.entity.user.repository.UserModelRepository;
import com.jbconnect.healthclinix.entity.user.repository.UserModelSpecification;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Component;
import org.springframework.data.domain.Pageable;

@Component
public class GetByFiltersAdapter implements GetByFiltersPort<Pageable, UserRequestDTO, UserResponseDTO> {

    private final UserMapper userMapper;

    private final UserModelRepository userModelRepository;

    public GetByFiltersAdapter(UserMapper userMapper, UserModelRepository userModelRepository) {
        this.userMapper = userMapper;
        this.userModelRepository = userModelRepository;
    }

    @Override
    public Page<UserResponseDTO> execute(Pageable pageable, UserRequestDTO userRequestDTO) {
        UserModel userModelMap = this.userMapper.userRequestToUserModel(userRequestDTO);

        return this.userModelRepository.findAll(UserModelSpecification.getByFilters(userModelMap), pageable)
                .map(this.userMapper::userModelToUserResponseDTO);
    }
}

File: backendkotlin/src/main/kotlin/com/jbconnect/healthclinix/healthclinix/entity/user/adapter/GetByFiltersAdapter.kt
package com.jbconnect.healthclinix.healthclinix.entity.user.adapter

import com.jbconnect.healthclinix.healthclinix.entity.user.dto.UserRequestDTO
import com.jbconnect.healthclinix.healthclinix.entity.user.dto.UserResponseDTO
import com.jbconnect.healthclinix.healthclinix.entity.user.mapper.UserMapper
import com.jbconnect.healthclinix.healthclinix.entity.user.model.UserModel
import com.jbconnect.healthclinix.healthclinix.entity.user.port.GetByFiltersPort
import com.jbconnect.healthclinix.healthclinix.entity.user.repository.UserModelRepository
import com.jbconnect.healthclinix.healthclinix.entity.user.repository.UserModelSpecification
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Component

@Component
class GetByFiltersAdapter(private val repository: UserModelRepository, private val mapper: UserMapper) :
    GetByFiltersPort<Pageable, UserRequestDTO, UserResponseDTO> {

    override fun execute(pageable: Pageable, userRequestDTO: UserRequestDTO): Page<UserResponseDTO> {
        val userModelMap: UserModel = this.mapper.userRequestToUserModel(userRequestDTO)
        return this.repository.findAll(UserModelSpecification.getByFilters(userModelMap), pageable)
            .map(this.mapper::userModelToUserResponseDTO)
    }

}